Tolerate linear solver failure in non-linear schemes

lhy11009 opened the pull request:
Pull Request Checklist. Please read and check each box with an X. Delete any part not applicable. Ask on the [forum](https://community.geodynamics.org/c/aspect) if you need help with any step.

*This is a tentative PR that I opened to first discuss it with the maintainers. The code change is merely for demonstrating the idea. See the comments below.*

### Before your first pull request:

* [x] I have read the guidelines in our [CONTRIBUTING.md](../blob/main/CONTRIBUTING.md) document.

### For all pull requests:

* [x] I have followed the [instructions for indenting my code](../blob/main/CONTRIBUTING.md#making-aspect-better).

### For new features/models or changes of existing features:

* [x] I have tested my new feature locally to ensure it is correct.
* [x] I have [created a testcase](https://aspect-documentation.readthedocs.io/en/latest/user/extending/testing/writing-tests.html) for the new feature/benchmark in the [tests/](../blob/main/tests/) directory.
* [x] I have added a changelog entry in the [doc/modules/changes](../blob/main/doc/modules/changes) directory that will inform other users of my change.

gassmoeller commented:
Hy Haoyuan, I think this is a very interesting idea. Is this mostly useful if you have strongly nonlinear material model properties that may lead to difficult / unrealistic results in the first nonlinear iterations?

Conceptually this is similar to choosing a coarse linear solver tolerance inside the nonlinear solver under the assumption that the nonlinear solver will iterate until its tolerance is reached. Something like this (changing the linear solver tolerance between nonlinear iterations) is possible in the Newton solver, where the linear solver doesnt need to reach a very tight tolerance anyway in order to improve the solution. However, your approach would work for any nonlinear solver.

We can check others opinions -- I would be curious to hear what @bangerth and @tjhei think about this -- but I wouldnt be opposed to adding a parameter like `Continue nonlinear solver loop after linear solver failure`. However I would suggest the following modifications, do you think they make sense?
- The parameter should not be limited to the cheap solver. If the cheap solver does not converge, and there are expensive iterations requested (by setting ` Maximum number of expensive Stokes solver steps` to larger than 0 or leaving it at the default 1000) then those number of expensive iterations should be performed next. If that parameter is set to 0, or the expensive solver does not converge either then we can check the same new parameter to decide if we should throw an exception and crash or not (see solver.cc:912).
- The last linear solve before exiting the nonlinear solver loop should succeed. I am not sure this is mathematically required, because if the nonlinear residual is small enough to fulfil the convergence criteria the Stokes residual should be small as well. But I would feel quite uncomfortable continuing with a Stokes solution that did not converge. Also many users set a maximum number of nonlinear iterations and continue even if the nonlinear solution is not converged. This would be dangerous if the linear solver didnt converge either. To check if the last Stokes solve converged may require storing the state (converged/not converged) of the last Stokes solution result somewhere (maybe we can work with the `SolverControl` objects we use inside `solve_stokes()`).
- (Maybe) we could require that the parameter `Nonlinear solver failure strategy` is **not** set to `continue with next timestep` if your new parameter is set. This would mean we can never get into the situation I described above (linear solver doesnt converge, nonlinear solver doesnt converge either, but we reach the maximum number of nonlinear iterations and continue anyway).

What do you think? 
Happy to hear the thoughts of others on the topic.

lhy11009 commented:
Hey Rene,

Yes—this is very useful in cases with strongly nonlinear material behavior during the first nonlinear iterations. In practice, it saves a significant amount of wall-clock time because the cheap iterations are much faster (even though I haven’t quantified this yet). I’ve used this scheme across both 2-D and 3-D models, as well as setups with slightly different configurations (e.g., resolution changes), and it has worked consistently well.

I think all of your suggestions make sense. In particular, it would be important to consider how this new parameter relates to “Maximum number of expensive Stokes solver steps” and “Nonlinear solver failure strategy.” Conceptually, I also agree that the final linear solve before exiting the nonlinear solver loop should still converge.

I’d be curious to hear what others think as well.

lhy11009 committed:
Tolerate linear solver failure in non-linear schemes

lhy11009 commented:
I tried to follow the lines of discussion we had, and now the implementation works partially. I also added a test using this new strategy. In this test, the nonlinear solver converges in one step after ~40 iterations, despite the linear solver reporting a failure.

However, I ran into an issue regarding the implementation.

Ideally, we would only modify the nonlinear solver so that it continues the iteration even if the linear solver fails. But I realized that this does not work in practice: if we only change how the nonlinear solver handles exceptions, the linear solver will still throw the exception before updating the residual, so the nonlinear solver cannot proceed correctly.

Instead, in this example I went to the lowest level of the linear solver, where the exception is first thrown, and modified it so that it continues from there.

This might not be the best design choice, so I am very open to suggestions on how this should be handled.

gassmoeller reviewed (commented):
I agree with your assessment, we need to indeed do this at the level of the linear solver. I think your general approach is correct, but I have some comments to the code (see below).

Also: could you model this approach after what we do with `nonlinear_failure_strategy`? I.e. use an `enum` type to allow more possible behaviors in the future (e.g. we could extend this at some point to cut the timestep if the linear solver fails).

Also please introduce a variable similar to `nonlinear_solver_failures` that keeps track of how often the linear solver failed, and then output that number at the end of the simulation like done in simulator/core.cc:1280.

gassmoeller reviewed (commented):
Yes, I think this is the right behavior.

I would like to ask for the following changes, just to improve the safety of the functionality: 

1. Add a variable `linear_solver_failures` in simulator.h like the variable `nonlinear_solver_failures`. Search for all places where `nonlinear_solver_failures` is used and let `linear_solver_failures` behave the same way. Add a warning at the end of the model run (like in core.cc:2329) if any linear solver failures occured and report how many. This will document for the user how many times the linear solver failed (and hopefully push them to investigate if too many failures happened). Also I noticed that `nonlinear_solver_failures` is not serialized into the checkpoint for restarts. We should do that (and do the same for `linear_solver_failures`). Can you add both to checkpoint_restart.cc:609?

2. Add a function to parameters.h similar to `is_defect_correction` but name it `solve_Stokes_iteratively` and check if any iterative Stokes solver scheme is selected. Then add something like the following in parameters.cc after reading the linear solver failure strategy:
```
if (linear_solver_failure_strategy == continue)
  AssertThrow(Parameters<dim>::solve_Stokes_iteratively(nonlinear_solver) == true, ExcMessage("You are not allowed to select the linear solver failure strategy 'continue' with a solver scheme that does not iterate the Stokes equations."));
```

This will not be quite as safe as I originally hoped for (checking if the last Stokes solver converged), but at least it will make sure no one can continue with a non-converged solution if the solver scheme does not at least try to iterate.

lhy11009 commented on source/simulator/solver/stokes_matrix_free_local_smoothing.cc:
```diff
@@ -1509,6 +1509,8 @@ namespace aspect
         // if the solver fails trigger the post stokes solver signal and throw an exception
         catch (const std::exception &exc)
           {
+            // ++linear_solver_failures;
```
This is related to counting the linear_solver_failure number. However, Adding this here ("GMG" solver) doesn't work because this function is not part of the simulator. The other location of the "AMG" solver doesn't have this problem.

gassmoeller replied:
The GMG solver has a workaround to do this, which is to make this into:
```
++sim.linear_solver_failures;
```
(the GMG solver stores a reference to the Simulator in `sim`)

lhy11009 replied:
Yes, this works!

lhy11009 commented:
> Yes, I think this is the right behavior.
> 
> I would like to ask for the following changes, just to improve the safety of the functionality:
> 
> 1. Add a variable `linear_solver_failures` in simulator.h like the variable `nonlinear_solver_failures`. Search for all places where `nonlinear_solver_failures` is used and let `linear_solver_failures` behave the same way. Add a warning at the end of the model run (like in core.cc:2329) if any linear solver failures occured and report how many. This will document for the user how many times the linear solver failed (and hopefully push them to investigate if too many failures happened). Also I noticed that `nonlinear_solver_failures` is not serialized into the checkpoint for restarts. We should do that (and do the same for `linear_solver_failures`). Can you add both to checkpoint_restart.cc:609?
> 2. Add a function to parameters.h similar to `is_defect_correction` but name it `solve_Stokes_iteratively` and check if any iterative Stokes solver scheme is selected. Then add something like the following in parameters.cc after reading the linear solver failure strategy:
> 
> ```
> if (linear_solver_failure_strategy == continue)
>   AssertThrow(Parameters<dim>::solve_Stokes_iteratively(nonlinear_solver) == true, ExcMessage("You are not allowed to select the linear solver failure strategy 'continue' with a solver scheme that does not iterate the Stokes equations."));
> ```
> 
> This will not be quite as safe as I originally hoped for (checking if the last Stokes solver converged), but at least it will make sure no one can continue with a non-converged solution if the solver scheme does not at least try to iterate.

I have addressed these points, except for the issue I commented on in the code.

gassmoeller commented on source/simulator/checkpoint_restart.cc:
```diff
@@ -750,6 +750,8 @@ namespace aspect
     ar &pre_refinement_step;
     ar &last_pressure_normalization_adjustment;
     ar &total_walltime_until_last_snapshot;
+    ar &nonlinear_solver_failures;
```
oh, so this wasnt serialized before? This is actually a bugfix.

gassmoeller reviewed (commented):
Nice! I left a comment on how to fix the one issue you reported.

Could you also:
- rebase to current main branch
- write a changelog entry that describes both the bugfix (serialization of nonlinear solver failures), and the new feature (optionally ignore linear solver failures).

lhy11009 commented:
@gassmoeller I rebased and added a changelog.

gassmoeller reviewed (commented):
Ok, the code looks all good, but you will have to update the parameter documentation by running `make parameters`, or by downloading the diff file from the indent tester.

Ready to merge if you update the parameter documentation and the testers pass.

lhy11009 commented:
It's updated, and it seems that all tests passed!

gassmoeller reviewed (approved):
Looks all good to go. Thanks!
